rust/src/io/reader.rs

```diff
@@ -71,15 +75,34 @@ pub async fn read_manifest(object_store: &ObjectStore, path: &Path) -> Result<Ma
         ));
     }
     let manifest_pos = LittleEndian::read_i64(&buf[buf.len() - 16..buf.len() - 8]) as usize;
-    assert!(
-        file_size - manifest_pos <= buf.len(),
-        "Assert file_size - manifest_pos <= buf.len(), but got file_size = {}, manifest_pos = {}, buf.len() = {}",
-        file_size,
-        manifest_pos,
-        buf.len()
-    );
-    let proto =
-        pb::Manifest::decode(&buf[buf.len() - (file_size - manifest_pos) + 4..buf.len() - 16])?;
+    let manifest_len = file_size - manifest_pos;
+
+    let buf: Bytes = if manifest_len <= buf.len() {
+        // The prefetch catpured the entire manifest. We just need to trim the buffer.
+        buf.slice(buf.len() - manifest_len..buf.len())
+    } else {
+        // The prefetch only captured part of the manifest. We need to make an
+        // additional range request to read the remainder.
+        let mut buf2: BytesMut = object_store
+            .inner
+            .get_range(
+                path,
+                Range {
+                    start: manifest_pos,
+                    end: file_size - PREFETCH_SIZE,
+                },
+            )
+            .await?
+            .into_iter()
+            .collect();
+        buf2.extend_from_slice(&buf);
+        buf2.freeze()
+    };
+
+    // Need to trim the magic number at end, and the non-manifest data at the beginning
+    let buf = buf.slice(4..buf.len() - 16);
```

Comment:
What does this `4` stand for?

Reply:
hmm well my comment ", and the non-manifest data at the beginning" isn't right / specific. I'm not sure yet; all I knew is that it was skipped previously. I guess it's the u32 message length here:

https://github.com/eto-ai/lance/blob/d3f6f6c31c909da22c0b7d486bbc35ec101f6079/rust/src/io/object_writer.rs#L68

Perhaps we should be reading that length and validating it?

Reply:
Added validation for that and a better comment 👍